### ISSUE TITLE
Don't include files included by corresponding header

### ISSUE THREAD

**Issue by Flamefire** [include-what-you-use/include-what-you-use#272]
If you have a `foo.cpp` with `foo.h` where `foo.h` already includes `bar.h` and of course `foo.cpp` includes `foo.h` first then it should not be required to also include `bar.h`

**Comment by kimgr** [comment]
This should already be happening. We need a small repro to be able to act on this.

**Comment by Flamefire** [comment]
Before I start searching for one: Is it possible that IWYU gets confused, if the "corresponding" header is not the first one included? I got a common "defines.h" that (by convention) must be included first in every cpp file (similar probably to the stdafx.h on MSVC) so the "corresponding" header will be the 2nd one.

**Comment by cdglove** [comment]
> This should already be happening. 

Not always I don't think. This really depends on what the foos and bars look like and how one feels about transitive includes. It's often not a good practise to depend on header including other headers for you, so we really need to see what these foos and bars look like.

**Comment by kimgr** [comment]
>  Is it possible that IWYU gets confused, if the "corresponding" header is not the first one included?

Yes, the asssociated header must be the first included, and match the name of the .cpp file.

>  I got a common "defines.h" that (by convention) must be included first in every cpp file

If it's really included in *every* cpp file, you could probably use the `--pch_in_code` switch to ignore it (we added that to cope with `stdafx.h`).

> Not always I don't think. This really depends on what the foos and bars look like [...]

It only happens for so-called associated headers, i.e. the A.h for A.cpp. The theory is that A.cpp will always include A.h, and they belong together, so there's no point in duplicating includes between them.

**Comment by Flamefire** [comment]
Well it is a rather large project with a couple of subprojects. One of them is even an external C project (library) so no it is no in ***ALL*** cpp files. In one of the projects it is `defines.h` in another it is `main.h` and in the C project there is no such thing.

Wouldn't it make sense, to extent the `pch_in_code` option so that it checks, if there is an additional header before the "corresponding" header and so detect the "corresponding" header also on 2nd position?

**Comment by kimgr** [comment]
OK, then you pass `--pch_in_code` to the projects that have it.

That's basically what the pch-in-code logic does, but there's no conditional evaluation -- if you've specified pch-in-code, the first header is always marked as a PCH (and evaluated according to `--prefix_header_includes`). If not, the first header is assumed to be the associated header.

**Comment by Flamefire** [comment]
It seems to associate the header even without the pch option. As I currently don't have time to condense the example I just attach what I have. The project is here: https://github.com/Return-To-The-Roots/s25client
In my example I use the latest master from the `liblobby` sub-repo. After the (pretty short and well-described) setup steps I run iwyu with: `.../wyu_tool.py -v -p . /home/<user>/git/s25client/liblobby/src/LobbyClient.cpp -- --verbose=5 2&> out.txt`

The output is here: https://gist.github.com/Flamefire/6a2cdd915af9eac1f408

Note the following:

> Marked /home/<user>/git/s25client/liblobby/src/LobbyClient.h as associated header of /home/<user>/git/s25client/liblobby/src/LobbyClient.cpp.
>...
>[ Exiting to  ] /home/<user>/git/s25client/liblobby/src/LobbyClient.h:26:36
[ #include    ] "LobbyServerList.h" (/home/<user>/git/s25client/liblobby/src/LobbyServerList.h)
Added an #include: /home/<user>/git/s25client/liblobby/src/LobbyClient.h -> "LobbyServerList.h"
>...
>/home/<user>/git/s25client/liblobby/src/LobbyClient.cpp should add these lines:
>...
#include "/home/<user>/git/s25client/liblobby/src/LobbyServerList.h"  // for LobbyServerList

So in short: associated header is found, include is found in associated header, still it suggests including it. Note that it also suggests including the header with its full path although they are in the same directory.

I can try to condense the example, but it might not be that easy. Maybe debugging or another output from iwyu already helps to identify the cause?

**Comment by Flamefire** [comment]
I was able to reproduce this with a really small example and also identify a possible cause:
[incTest.zip](https://github.com/include-what-you-use/include-what-you-use/files/142703/incTest.zip)

If you check the example in-tree with `include-what-you-use User.cpp` then the output is correct (no suggested changes). But if you compile from another folder (e.g a `build`-subfolder)  or specify the path absolute then it suggests adding the include.

So it seems that the bug always occurs, when there is a path in front of the file to be checked.

**Comment by Flamefire** [comment]
This behaviour is actually caused by this bug: #271 

The directory of a file should always considered as an implicit include path for this file. More info in the bug itself.
